### DISCUSSION TITLE
Feedback: new routing and layout make svelte kit barely usable on linux / terminal

### DISCUSSION THREAD

**Discussion by mef** [sveltejs/kit#7206] [General]
Hi, 

Thanks for Svelte and for SvelteKit. Using these in the past 1.5 years simplified by web development activity, enabled me to deliver quality software with super easy way to maintain and modify the app.

I recently upgraded past @1.0.0-next.405 and went through the migration and adaptation to new routing system and new layouts.

I've read a fair share of some of the discussions content about both of these, and feel the need to add the perspective of a linux user not using the same tooling as the majority, as the new design forces me to consider other options than SvelteKit, regretably.

## A bit of background

As a user of svelteKit, I wish to:

* use my text editor of choice, which is not `VSCode`,
* use git from the terminal (`terminator` on linux) while keep using the same git workflow I am used to with svelteKit as with all other technology projects I work on,
* enjoy a "beautiful development experience", as advertised in [svelteKit project's home page](https://kit.svelte.dev/).

The recent changes seem to prevent these expectations from being respected, and force me to consider other solutions.

## Issues

Summary:

* Replacing named layouts by layout groups introduces some un-necessary boilerplate, likely affecting many svelteKit projects
* The layout groups' files naming convention **prevents using git in the command line** as before, and as in any other project
* The files naming convention introduced in the new routing system prevents using git in the command line as before, and as in any other project
* All files have the same name, when this really doesn't seem necessary at all, given the mandatory prefix `+`

Here are details about the issues encountered. The proposed solutions are just naive ideas on the top of my head, without much knowledge about Svelte kit's internals, and I understand that they might not be acceptable or suitable.

#### Situation

A website contains pages with two distinct layouts.

For example:

* a default layout with a classic blog structure
* a visual layout with showing full page interactive slides. No navbar, footer.

#### Previous situation

```
src/routes/
├ __layout.svelte
├ __layout-visual.svelte
├ index.svelte
├ about.svelte
├ knowledge-base/
│ ├ topic-1/
│ │ ├ index.svelte
│ │ └ topic1-slides@visual.svelte
│ ├ topic-2/
│ │ ├ index.svelte
│ │ └ topic2-slides@visual.svelte
```

#### Current situation

Because the visual layout should not inherit from the default layout, it is necessary to wrap the whole website in an layout group, so that a `root` layout only contains `<slot></slot>` (this corresponds to `use case 2` in the layout migration guide #6174).

```
src/routes/
├ +layout.svelte
├ (site)
│ ├ +layout.svelte
│ ├ +page.svelte
│ ├ about
│ │ └ +page.svelte
│ ├ knowledge-base/
│ │ ├ +layout@.svelte
│ │ ├ topic-1/
│ │ │ └ +page.svelte
│ │ ├ topic-1-slides/
│ │ │ └ +page@knowledge-base.svelte
│ │ ├ topic-2/
│ │ │ └ +page.svelte
│ │ ├ topic-2-slides/
│ │ │ └ +page@knowledge-base.svelte
```

#### Why is it a big deal?

An extra level of nesting isn't the end of the world. However, in this specific case, this is problematic because of the file naming convention, preventing to navigate through the directory tree naturally in a terminal, using <kbd>Tab</kbd> for path completion (c.f. next issue).

#### Proposed solution

Include a layout reset mechanism not requiring to wrap the whole website under an extra directory. 

### Layout groups: route files not usable in the terminal in git commands anymore

Each time a file inside `src/routes` is modified, it is not possible anymore to use usual terminal features to stage and commit the file in git.

#### Workflow

1. type `git stat` + <kbd>Tab</kbd> + <kbd>Enter</kbd> to show the git status
2. double-click on the name of modified file to commit. The whole file path and name is selected
3. type `git add `
4. Click with the middle-mouse button to paste the selected file path and name + <kbd>Enter</kbd> (a very common and time-saving operation in Linux)

#### Previous situation

Result: the desired file is staged for commit

#### Current situation

Two issues:

1. step 2 of the workflow does not select the whole file name + path anymore, but only a subsection until next parenthesis sign(s). 
2. Assuming that the complete file name + path has been input, step 4 fails with the following error message: 

> bash: syntax error near unexpected token `('

#### Workflow adapted to the new layout groups syntax

1. type `git stat` + <kbd>Tab</kbd> + <kbd>Enter</kbd> to show the git status
2. click-and drag on the name of modified file to commit in order to select the whole file path and name is selected. This requires more precision in pointing and takes more time.
3. type `git add "` (notice the extra double-quote sign)
4. Click with the middle-mouse button to paste the selected file path and name
5. type `"` + <kbd>Enter</kbd>

#### Proposed solution

Either rollback to previous layout system, or switch to a terminal-friendly naming convention.

Side-note: terminal's auto-completion is also broken, e.g. with command `ls`. path needs to be escaped with a double-quote character in order to navigate through paths containing parenthesis. Affects, `terminator`, `xterm` and `urxvt` on Debian at least.

### New routing system: route files not usable in the terminal in git commands anymore

Similarly as in the previous issue, the `+` character prevents a double-click to select the whole file path + name from a `git status` output.

The stated reason for using this prefix is to simplify a regex in place to match the files:

> the regex we have today:
> 
>     /(?:(?:^_|\/_)|(?:^\.|\/\.)(?!well-known))/
> 
> the regex we'll have tomorrow:
> 
>     /[\/]\+/
> 
> the difference is whether it's inclusive or exclusive.

Source: [this comment](https://github.com/sveltejs/kit/discussions/5748#discussioncomment-3282952).

What is the problem with the longer regex? Once it is in place and working, it is not likely to require maintenance. Is it causing any significant performance issues?

#### Proposed solution

It is again a very naive approach, not knowing about constraints, but I thought I would present it, perhaps to hint on a solution involving _more simplicity_:

As pages, endpoints and server endpoints are the norm, and colocated files (component code stored inside the routing directory) are the exception:

* do not require any prefix for the pages, endpoints and server endpoints
* require a prefix to the other colocated files, which should _not_ generate a route.
* For actual route files, if the regex is really an issue, match on the file extension / suffix:
  * `.svelte` --> route file
  * `.server.js` --> server endpoint (not sure of the current terminology for this type of files)
  * `.js` --> endpoint

### all files have the same name

This has been debated a lot, but I did not see feedback from the perspective of developers not using `vsCode`, and for which a workaround to display directory name instead of file name is not available.

The `+` prefix and the forced `page` file name really look redundant.

Question: If a `+` prefix is required to the file names, then why not allowing them to be called anything rather than `page`? Routes would then look something like this:

```
src/routes/
├ home/
│ ├ +home.js
│ └ +home.svelte
├ about/
│ ├ +about.server.js
│ ├ +about.js
│ └ +about.svelte
```

## Questions

Thanks for your patience, and thanks again to the Svelte Kit creators and contributors for the wonderful work you have done.

* Can / should any of the issues reported above be considered as bug and be reported as such? Not being able to smoothly use the project with git on the terminal looks like a blocker to me, but I'm not sure everyone sees it the same way.
* Assuming the recent refactoring of both the routing and layout systems are final and will not be fixed to accommodate the needs of a broader range of users, is there any alternative svelte-based framework which someone may recommend?

Edits:

* remove duplicate discussion title
* add questions after reading that SvelteKit is now in Release Candidate phase

**Comment by kvetoslavnovak** [comment]
I have no problem with files names or “+”.
But I believe many many devs would welcome to substitute “+” or “page” keywords with their own using some kind of configuration possibility. E.g.  [7176](https://github.com/sveltejs/kit/discussions/7176)
Svelte has been well known as dev friendly after all.

**Reply by machak** [reply]
@mef agreed, pretty much unusable, but I don't think it will be changed back unfortunately. 
Your options are, (if using MPA at least) is to use Astro + Svelte or Routify plugin ( https://www.routify.dev/ ).

see also discussion at:
https://github.com/sveltejs/kit/discussions/6005 

I moved all of my site (content related) projects to Astro + Svelte and it is much better dev experience compared to SK. Astro comes short if you need SSR + SPA capabilities.

**Comment by raythurnvoid** [comment]
I think the new layout system is a significant improvement, it clearly separates each architectural piece; allows you to understand what's running only on the server and what will run in the client as well and it also allows you to create complex layout inheritance without encountering occasional issues (too hard and long to explain, but I had one of these and the new layout system fixed it). 

It may be a little bit more complex than before, but sveltekit is growing, and it allows us to do more every day while still being the easiest solution out there. Fairly speaking, I think that Astro is much harder, especially when you want to make different components/pages communicate between them, which is kind of the price to pay if you want to do "islands" and multiple frameworks in the same project.

**Comment by ejsmith** [comment]
I love Svelte and I love SvelteKit generally, but every single file being called +page.svelte and not being able to arbitrarily pick a layout for a file that isn't based on folder hierarchy is a really bad developer experience. I keep having to squint at my folder structure to try and figure out what file I'm working on and every single tab in my editor is the same file name is terrible. There has got to be a better way.

**Comment by cwoz117** [comment]
The big problem for me is vim considers the + to be a token when you open a file that starts with it, so you need to be in the source directory. 
`vim routes/+page.svelte` works, but `vim +page.svelte` does not.

Further, using hidden URI routes, (directory names wrapped in brackets) can't be pathed without wrapping the path in quotes: `vim "routes/(auth)/login/+page.svelte"` so even though staying in `src/` is demanded as per above, you're penalized for it if you want to use a Svelte feature.

I'm digging this convention for route layouts, but we DO need some sort of resolution to bash navigation.

**Reply by fehnomenal** [reply]
What about `vim ./+page.svelte`? I guess not ideal but maybe a acceptable workaround?

**Comment by domo84** [comment]
As much as I enjoy creating websites with Svelte Kit and I really think the file based routing is a great idea in theory. But the naming conventions are so bad I couldn't believe it at first. Been using Linux and CLI for 20 something years and I have never even once come remotely close to something as bat shit crazy as using +()@[] in filenames. I'm not even exaggerating when I tell you it breaks my workflow once an hour.

Just now, which brought me here, inside vim: `!git rm %`:
```
:!git rm src/routes/(private)/file/[...slug]/+page.svelte
/bin/bash: -c: line 1: syntax error near unexpected token `('
/bin/bash: -c: line 1: `git rm src/routes/(private)/file/[...slug]/+page.svelte'
```

**Comment by justuswilhelm** [comment]
This can even happen when you use fish shell on macOS. One particular issue I have with layout groups (using `()`), is that they are interpreted by the fish shell as command interpolation, similar to `$()` in bash et al. Using unconventional syntax for path specification can have so many unpredictable ripple effects.

I suspect that there is a bit of a Microsoft VS Code bias here, which can be seen in other tools surrounding the svelte "ecosystem" as well that offer first class VS Code support. Don't get me wrong. SvelteKit is preferable for me over next.js, and Svelte over React, but some developers might be better served with getting picked up from where they are right now. When using fully integrated development tools, they hide away all the dirty OS and shell behavior from you quite well.

It's inconvenient, but not bad enough to make me stop using SvelteKit. The ship for a change to folder structure has probably sailed now. A configuration flag for path syntax would mean that I will stray from the happy path, and it might break in weird ways, as experience tells me.
